pennylane_lightning/lightning_qubit.py

```diff
@@ -354,40 +408,61 @@ def apply_lightning(self, state, operations):
+                if batch_size_op
+                else state[np.newaxis, ...]
+            )
 
-        # Skip over identity operations instead of performing
-        # matrix multiplication with the identity.
-        skipped_ops = ["Identity"]
+        for batch_ind, state in enumerate(batched_states):
+            state_vector = np.ravel(state)
 
-        for o in operations:
-            if o.base_name in skipped_ops:
-                continue
-            name = o.name.split(".")[0]  # The split is because inverse gates have .inv appended
-            method = getattr(sim, name, None)
-
-            wires = self.wires.indices(o.wires)
-            if method is None:
-                # Inverse can be set to False since qml.matrix(o) is already in inverted form
-                method = getattr(sim, "applyMatrix")
-                try:
-                    method(qml.matrix(o), wires, False)
-                except AttributeError:  # pragma: no cover
-                    # To support older versions of PL
-                    method(o.matrix, wires, False)
+            if self.use_csingle:
+                sim = StateVectorC64(state_vector)  # use_csingle
             else:
-                inv = o.inverse
-                param = o.parameters
-                method(wires, inv, param)
-
-        return np.reshape(state_vector, state.shape)
+                # self.C_DTYPE is np.complex128 by default
+                sim = StateVectorC128(state_vector)
+
+            # Skip over identity operations instead of performing
+            # matrix multiplication with the identity.
+            skipped_ops = ["Identity"]
+            for o in operations:
+                if o.base_name in skipped_ops:
+                    continue
+                name = o.name.split(".")[0]  # The split is because inverse gates have .inv appended
+                method = getattr(sim, name, None)
+
+                wires = self.wires.indices(o.wires)
+                if method is None:
+                    # Inverse can be set to False since qml.matrix(o) is already in inverted form
+                    method = getattr(sim, "applyMatrix")
+
+                    try:
+                        _matrix = qml.matrix(o)[batch_ind] if o.batch_size else qml.matrix(o)
+                    except AttributeError:  # pragma: no cover
+                        # To support older versions of PL
+                        _matrix = o.matrix()[batch_ind] if o.batch_size else o.matrix()
+
+                    method(_matrix, wires, False)
+                else:
+                    inv = o.inverse
+                    param = np.transpose(np.array(o.parameters, copy=False))
+                    param = param[batch_ind] if o.batch_size else param
+                    method(wires, inv, param)
+            batched_states[batch_ind, ...] = np.reshape(state_vector, state.shape)
+        batched_states = batched_states if batch_size_op else batched_states[0, ...]
```

Comment:
As the whole differentiation between broadcasted states and non-broadcasted states happens within this function, would it maybe be better to 
1. not artificially create a "dummy" batch dimension and extract the first state at the end, but
2. outsource the application code itself and call it once for non-broadcasted cases and repeatedly for broadcasted cases?

Reply:
I thought about that, but this would increase the maintenance burden and ask for a more significant refactor.
More work and no apparent increase in performance.

Reply:
Okay, fair enough. I think I personally would find the other structure a bit easier to digest, but that's certainly a personal preference. The benefit of course would not be in performance, that's right :)